### DISCUSSION TITLE
Mania judgements placed in subfolders then imported to lazer are not converted properly

### DISCUSSION THREAD

**Discussion by vtaggerungv** [ppy/osu#16403] [Q&A] [answered]
This issue started for me personally when I began testing lazer, and immediately noticed that the judgement text ("perfect", "great", "miss", etc) in my imported skin was about twice the size as it was in stable. I tested other skins but the ones I tested didn't seem to have the same issue mine had, which was  very odd. That was until I checked the folder of my stable skin and realized that the judgements were placed in a subfolder and read via several lines in the skin.ini. After some testing with another skin I can confirm that when judgement sprites are pulled from a subfolder they are displayed in lazer at HALF size. The difference being that in my skin, the sprites I was using were appended with "@ 2x" in the name as well as actually doubled in size, which seems to actually double the size from normal.

Sprites not in subfolder (looks normal!)
![spk lazer no subfolder](https://user-images.githubusercontent.com/53627663/148862252-b090cac2-7e2a-4761-96a6-2bd6ce1d0a47.png)

Sprites placed in subfolder at 1x size (appears half size!)
![spk lazer subfolder 1x](https://user-images.githubusercontent.com/53627663/148862255-4514d5ad-0cf2-4092-8217-e7fea66652bc.png)

Sprites placed in subfolder at 2x size (appears twice the size as normal!)
![spk lazer subfolder 2x](https://user-images.githubusercontent.com/53627663/148862257-627a68cd-41c5-4290-abaf-31c8a4829f08.png)

Here are links to each respective skin:
No subfolder: https://www.dropbox.com/s/vw44ibunli2emkd/etienne%20skin%20no%20subfolders.osk?dl=0
Subfolder 1x: https://www.dropbox.com/s/wfnkqlor9wupodp/etienne%20skin%20subfolders%20%401x.osk?dl=0
Subfolder 2x: https://www.dropbox.com/s/ckg7n62zyly0tvu/etienne%20skin%20subfolders%20%402x.osk?dl=0

**Comment by bdach** [comment]
so a few things here:

the `@1x` skin displaying judgements at half size is expected. in that one you've replaced `@2x` assets with the normal sized ones. the game halves the size of `@2x` sprites before displaying them on the screen, to, well, make sure their final size is what is expected.

the `@2x` case displaying judgement at double size is not expected, however. that is caused by the following lines in your `skin.ini`:

```
Hit0: judge\mania-hit0@2x
Hit50: judge\mania-hit50@2x
Hit100: judge\mania-hit100@2x
Hit200: judge\mania-hit200@2x
Hit300: judge\mania-hit300@2x
Hit300g: judge\mania-hit300g@2x
```

lazer is not expecting component names to already contain `@2x` in their names, it assumes that every component name it receives does not have a scale suffix. so when the above is specified, it first tries to find what it *thinks* should be the HQ sprite at `judge\mania-hit0@2x@2x`, fails, and then hits the `judge\mania-hit0@2x` sprite, which is *suffixed* with `@2x`, but the game is expecting a normal-sized sprite at this point and so displays the `@2x` sprite as if it were a `@1x` sprite.

not sure what stable does to avoid this, but the dumb lazer fix would go something like this:

```diff
diff --git a/osu.Game/Skinning/LegacySkin.cs b/osu.Game/Skinning/LegacySkin.cs
index e677e2c01b..5a3299bd0e 100644
--- a/osu.Game/Skinning/LegacySkin.cs
+++ b/osu.Game/Skinning/LegacySkin.cs
@@ -474,13 +474,15 @@ public override Texture GetTexture(string componentName, WrapMode wrapModeS, Wra
         {
             foreach (string name in getFallbackNames(componentName))
             {
+                string strippedName = name.Replace("@2x", "");
+
                 float ratio = 2;
-                var texture = Textures?.Get($"{name}@2x", wrapModeS, wrapModeT);
+                var texture = Textures?.Get($"{strippedName}@2x", wrapModeS, wrapModeT);
 
                 if (texture == null)
                 {
                     ratio = 1;
-                    texture = Textures?.Get(name, wrapModeS, wrapModeT);
+                    texture = Textures?.Get(strippedName, wrapModeS, wrapModeT);
                 }
 
                 if (texture == null)
```

**Reply by peppy** [reply]
Unless this is a common thing that comes up in many skins, I'd say the user should just fix their `skin.ini`.

@vtaggerungv was this skin one you found online, or crafted yourself?

**Reply by vtaggerungv** [reply]
So out of the 5 skins I tested, the only one that was structured like this was the one that I used as my main skin (which I found online). At least now that I understand what is causing the problem, it's not a big deal. However, I can imagine some subset of players converting their skin to lazer and being confused the way that I was when their judgements are comically large. Perhaps when importing a skin that's like this, the notification tab will throw some warning about this?

**Reply by peppy** [reply]
Might be fine to apply the patch as above.

**Reply by bdach** [reply]
PR'd as https://github.com/ppy/osu/pull/16477

**Comment by bdach** [comment, marked as answer]
Should be resolved via https://github.com/ppy/osu/pull/16477
